add Dynamic object serializer that allows to convert kotlin data to dynamic

exaV opened the pull request:
Supports maps with primitive keys, lists, and classes

exaV commented:
Implements https://github.com/Kotlin/kotlinx.serialization/issues/116

exaV commented:
I am mainly looking for feedback. There are at least 3 TODOs in the code for which I am looking for feedback.
* Should we serialize nulls by omitting the key and value in the output. Or should we add the key with `null` as the value
* Serialdescriptor.getElementName() sometimes returns an int if the datastructure is a collection. Can I use that or is it just an implementation detail?
* how does WriteMode.POLY_OBJ differ from WriteMode.OBJ and how should we serialize and object in that mode?

sandwwraith commented on runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt:
```diff
@@ -0,0 +1,146 @@
+    data class Data(val a: Int)
+
+    @Serializable
+    open class DataWrapper(open val s: String, val d: Data?)
```
Same question about `open`

exaV replied:
The test shows that the names are not mangled after serialization.

sandwwraith commented on runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt:
```diff
@@ -0,0 +1,146 @@
+        val data2 = WithChar('c')
+        val serialized2 = DynamicObjectSerializer().serialize(data2)
+        assertEquals(Json.stringify(data2), JSON.stringify(serialized2), "should serialize characters as strings")
+        assertEquals(data2.a.toString(), serialized2.a) // FIXME why is toString needed here, both should be char?
```
Probably, because natively chars are represented as ints in Kotlin/JS. You can leave this `.toString()`

sandwwraith commented on runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt:
```diff
@@ -0,0 +1,146 @@
+        assertEquals(Json.stringify(data), JSON.stringify(serialized))
+        assertEquals(data.a, serialized.a)
+        assertNotNull(serialized.list)
+        assertEquals(data.list[0], serialized.list[0])
```
Why not make a loop?

sandwwraith commented on runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt:
```diff
@@ -0,0 +1,146 @@
+    fun complexMapWrapperTest() {
+        val data = ComplexMapWrapper(mapOf("key1" to Data(1), "key2" to Data(2)))
+        val serialized = DynamicObjectSerializer().serialize(data)
+        assertEquals(Json.stringify(data), JSON.stringify(serialized))
```
You can reduce copy-paste by extracting these 3 lines into a function similar to `assertStringFormAndRestored` (e.g. `assertDynamicForm`)

pdvrieze commented on docs/runtime_usage.md:
```diff
@@ -220,5 +220,25 @@ val dyn = js("""{s:"foo", d:{a:42}}""")
+
+val wrapper = DataWrapper("foo", Data(42))
+JSON.stringify(wrapper) // {"s_dsrefg$_0":"foo","d":{"a":42}}
+val plainJS: dynamic = DynamicObjectSerializer().serialize(wrapper)
```
In this case are you mainly trying to use serialization to get a representation that removes the name mangling for properties (and thereby is usable with plain javascript)? Any reason not to use `@JsName` instead?

sandwwraith replied:
It also affects data structures like lists as well

exaV commented on docs/runtime_usage.md:
```diff
@@ -220,5 +220,25 @@ val dyn = js("""{s:"foo", d:{a:42}}""")
+data class Data(val a: Int)
+
+@Serializable
+open class DataWrapper(open val s: String, val d: Data?)
```
I thought it was important to show how we are unaffected by name mangling here since it is one of the use cases for DynamicObjectSerializer. 
We could also just mention it in the text above and leave out the `open val` from the example if you prefer.

exaV commented:
Thanks for having a look :)

I have one TODO left in the code on line 115 in DynamicObjectSerializer.kt. I am not sure what WriteMode.POLY_OBJ is about.

exaV commented:
The implementation encodes maps as JS objects `"{}"` and thus throws an exception when a key is not String or Int. 
We could convert to [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) instead, which has been supported by major browsers for some time, with an opt-in parameter for DynmicObjectSerializer. However I have not seen any support for Map in all of Kotlin/js so I am a bit hesitant to just add it.

sandwwraith commented:
@exaV 
POLY_OBJ is basically a mode which is used by useArrayPolymorphism flag (see more here: https://github.com/kotlin/kotlinx.serialization/blob/c4acfb69b8/docs/polymorphism.md#a-bit-of-customizing). So leaving js(BEGIN_LIST.toString() + END_LIST) here will likely get things done.

> The implementation encodes maps as JS objects "{}" and thus throws an exception when a key is not String or Int.

I think this is acceptable behaviour; use case with javascript `Map` is quite rare.

I see you've fixed most of my remarks, but there are some of them left. Do you have further questions?

ankushg commented on runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt:
```diff
@@ -0,0 +1,191 @@
+    private inline fun <reified T : Any> assertDynamicForm(
+        data: T,
+        serializer: SerializationStrategy<T>? = null,
+        noinline assertions: ((T, dynamic) -> Unit)? = null
```
Should `assertions` be allowed to be null? It seems like there are currently quite a few test cases that don't assert anything regarding the input/output, but that is likely important?

ankushg commented on runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt:
```diff
@@ -0,0 +1,191 @@
+//        assertDynamicForm(Color.GREEN, serializer = CommonEnumSerializer("Color"))
+        assertDynamicForm(EnumWrapper(Color.RED)) { data, serialized ->
+
+        }
```
```suggestion
        assertDynamicForm(EnumWrapper(Color.RED))
```

Doesn't seem like you have any custom assertions, so we might as well omit it?

ankushg commented on runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt:
```diff
@@ -0,0 +1,191 @@
+
+    @Test
+    fun enums() {
+//        assertDynamicForm(Color.GREEN, serializer = CommonEnumSerializer("Color"))
```
Should this be uncommented? Are custom serializers for enums supported?

exaV replied:
it was not supported before, I added support for it in the last commit.

ankushg commented on runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt:
```diff
@@ -0,0 +1,191 @@
+
+    enum class Color {
+        RED, GREEN
+    }
```
Should this class be moved to the top, with the rest of the classes that you're testing?

ankushg commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,130 @@
+        WriteMode.OBJ -> js(BEGIN_OBJ.toString() + END_OBJ)
+        WriteMode.MAP -> js(BEGIN_OBJ.toString() + END_OBJ)
+        WriteMode.LIST -> js(BEGIN_LIST.toString() + END_LIST)
+        WriteMode.POLY_OBJ -> TODO()
```
The reason we don't just use `begin`/`end` from the `WriteMode` itself (and essentially redeclare them here) is because the compiler doesn't treat `begin`/`end` as string constants, correct?

exaV commented:
Hi all, thanks for the reviews. Special thanks to @sandwwraith, you actually left a very good review the first time around and answered all my questions. I didn't see the answers at first because they were collapsed and marked as outdated by Github, I will definitely check "outdated" comments more carefully in the future.

I (think) I accounted for all the suggestions made. Please let me know if there is anything we should change.

Compared to the first round of reviews, we now:
* support enums
* support singular values e.g. a primitive Int (for consistency's sake)
* reflect the tree like structure of the serialization output in our naming

ankushg commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,167 @@
+            return obj
+        }
+        if (strategy.descriptor.kind is UnionKind.ENUM_KIND) {
+            return obj.unsafeCast<Enum<*>>().name
```
Does this still work if we have a custom serializer for the enum that _doesn't_ use `name`?

exaV replied:
The if block will match any singular enum e.g. `DynamicObjectSerializer.serialize(Color.RED)`.

I am not sure what you are getting at. Are you worried about about a use-case where someone would want their enums to be serialised to a different name?

```
enum class Color {
@SerialName("red")
RED
}
```

The code above would not consider the SerialName annotation, or a custom serialiser if there were one. I think we would have to introduce a `SingularValueEncoder` to handle the case correctly.

exaV replied:
The if block will match any singular enum e.g. `DynamicObjectSerializer.serialize(Color.RED)`.

I am not sure what you are getting at. Are you worried about about a use-case where someone would want their enums to be serialised to a different name?

```
enum class Color {
@SerialName("red")
RED
}
```

The code above would not consider the SerialName annotation, or a custom serialiser if there were one. I think we would have to introduce a separate `DynamicValueEncoder` class to handle the case correctly, because `DynamicObjectEncoder` only deals with classes and collections.

ankushg replied:
> Are you worried about about a use-case where someone would want their enums to be serialised to a different name?

Yeah that's exactly it. 

I know several folks (including myself) who use `@SerialName`, `@SerialId`, or custom `Serializer` implementations for enum classes (and of course other classes) since `ordinal`/`name` might not be consistent across codebases that still need to communicate with each other.

sandwwraith replied:
It would be nice to support enums with `@SerialName` here despite singular customized enums are rare case —because it is already supported in `encodeEnum` for wrapped enums. I've not tested it myself (yet), but i believe that replacing this line with `strategy.descriptor.getElementName(obj.unsafeCast<Enum<*>>().ordinal)` would be sufficient — since custom serial names are stored in descriptor.

ankushg commented on runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt:
```diff
@@ -0,0 +1,205 @@
+    data class EnumWrapper(val e: Color)
+
+    enum class Color {
+        RED, GREEN
```
Can we add some `@SerialId` and `@SerialName` test cases to make sure we cover those situations, especially with enums where it's easy to accidentally use `name`?

ankushg commented on runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt:
```diff
@@ -0,0 +1,205 @@
+    @Test
+    fun enums() {
+        assertDynamicForm(EnumWrapper(Color.RED))
+        assertDynamicForm(Color.GREEN, serializer = CommonEnumSerializer("Color"))
```
The default `CommonEnumSerializer` seems to just serialize using `name`, so I don't think this test is a good indicator for using custom enum serializers.

Can we replace this test with a serializer that behaves differently from the default behavior?

Maybe using something like 

```kt
val serializer = CommonEnumSerializer(
  serialName, 
  enumValues(), 
  enumValues<Color>()
    .map { "${it.name}_${it.ordinal}" }
    .toTypedArray()
)
```

exaV replied:
The test case may have been confusing, I only wanted to test the serialization of a singular enum. I will add `@Serializable` to the Color enum and remove the custom one here.

We can also add an additional test with custom serialisers.

ankushg commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,167 @@
+            currentName = if (elementDescriptor.kind is PrimitiveKind) {
+                value.toString()
+            } else if (elementDescriptor.kind is UnionKind.ENUM_KIND && value is Enum<*>) {
+                value.name
```
Should we honor custom serializers or `@SerialName`/`@SerialId` annotations for enums when being used as map keys?

exaV replied:
We now have two cases where we (naively) serialize Enum as `Enum.name`. We will have to again resort to something like `DynamicValueEncoder` to do it correctly.

Also if we were supporting custom serialisers here, we would have to support them for all types  iff the serialiser returns a legal javascript object key (that is String or Int).

ankushg replied:
> Also if we were supporting custom serialisers here, we would have to support them for all types iff the serialiser returns a legal javascript object key (that is String or Int).

Yeah that's fair. I think the most important thing is to be clear about what the limitations are.

sandwwraith replied:
You can apply here the same technique that I described in previous comment

sandwwraith replied:
However, there's a bigger problem with that: `encodeValue` is a fallback method, i.e. if map key is `Char` or `Enum`, `encodeChar` and `encodeEnum` are called, respectively — and this method won't be called. Seems like you have two options: 1. Move 'elementIsMapKey' check to a separate function and call it in the beginning of `encodeChar/Enum` 2. Move check to a `encodeElement` function (check itself does not require actual value, only descriptor) and then delegate to a `encodeValue` manually (e.g. `override fun encodeEnum(enumDescription: SerialDescriptor, ordinal: Int) = encodeValue(enumDescription.getElementName(ordinal))`

exaV replied:
Thanks for the suggestions, they helped to arrive a somewhat elegant solution.
We delegate writing operations to `encodeValue`. We also keep track of `currentElementIsMapKey` and use it in `beginStructure` to throw an exception and in  `encodeNull` to write null as the map key if necessary.

exaV replied:
We now support `@Serialname` on enums, both for map keys and for singular enums (and of course for regular properties).

exaV replied:
I also documented what is currently not supported which is 
```
 * Limitations:
 * * Map keys must be of primitive or enum type
 * * Custom serializers are not supported
 * * Enums are serialized as the value of `@SerialName` if present or their name, in that order. Ordinals are not supported.
 * * There is no analogue to DynamicObjectParsers' useArrayPolymorphism flag
```

exaV replied:
we now support custom serialisers everywhere :)

ankushg commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,167 @@
+        val newMode = selectMode(desc)
+
+        if (result === NoOutputMark) {
+
```
```suggestion
```

exaV commented:
Is there any way to support the annotation `@JsName` in a similar way to `@SerialName`? 

In existing code where I'd like to use `DynamicObjectSerializer` is already using `@SerialName` for interop. Unfortunately that annotation currently has no effect on this serializer. 

IMO the implementation should use `JsName` if `SerialName` is not used and some config property is enabled e.g. `useJsName`

ankushg commented:
I agree that it would be nice to factor `@JsName` into `dynamic` serialization, but maybe adding new serialization behavior/flags is out of scope for this PR?

I think it's probably best to focus this PR on serializing with the already-standard `@SerialId`/`@SerialName`/`Serializer` interactions, and make a separate PR to propose a new `@JsName` behavior

sandwwraith commented:
I'm not sure if `@JsName` annotation can be obtained at runtime in any way; afair, it is a compile-time annotation.

sandwwraith commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    override fun encodeNull() {
+        if (encodeDefaults &&
```
You don't need this check since `shouldEncodeElementDefault` is called by serializer _before_ any `encode` calls. So you should just override it: `override fun shouldEncodeElementDefault(desc: SerialDescriptor, index: Int): Boolean = encodeDefaults`

exaV replied:
Are you sure? Maybe I am doing something wrong but the shouldEncodeElementDefault method is not called at all if I omit the call here.
Do I need to rebase first?

sandwwraith replied:
No, it's quite old functionality. Maybe you confused `class X(val s: String? = null)` vs `class Y(val s: String?)` – in case of Y, `shouldEncodeElementDefault` indeed would not be called. If you still want to drop nulls regardless whether they are optional or not, (which may confuse `DynamicObjectParser`, which tries to make the difference between property being `null` and missing (`undefined`)), then you indeed should apply custom logic here (which basically would be 'do not write null if encodeDefaults=true)

exaV replied:
Thanks for clearing things up. I now understand that the intended behaviour of `shouldEncodeElementDefault` is to omit properties that *exactly* have the default value. Which totally makes sense given its name.

IMO the intended behaviour of `shouldEncodeElementDefault` is almost never what I would want from `DynamicObjectSerializer`.

The original problem I wanted to solve was indeed to differentiate between undefined and null. I will add an optional property `omitNullProperties = false` and document it accordingly. We could still leave the optional `shouldEncodeElementDefault = true` and change the implementation to what you suggested above.

sandwwraith commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    fun exitNode() {
+        current.parent
```
unused statement?

sandwwraith commented on runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt:
```diff
@@ -0,0 +1,338 @@
+
+        val serialized = DynamicObjectSerializer(encodeDefaults = true).serialize(data)
+        assertNull(serialized.d)
+        assertFalse(js("""Object.keys(serialized).includes("d")"""), "should omit null properties")
```
Idea of 'omitting defaults' is that we should omit _default property_ value, not _null_. Since the `DataWrapper` has declaration `val d: Data? = Data(1)`, we should omit d only if it is equal to `Data(1)`. That's automatically detected by generated serializer and then `shouldEncodeElementDefault` is called.

sandwwraith commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,180 @@
+        if (currentElementIsMapKey) {
+            currentName = null
+        } else {
+            if (shouldEncodeElementDefault(currentDescriptor, current.index) &&
```
If you want the converter to omit exactly `null`s, not all the default values (although I'm not sure if this is reliable in terms of parsing data back with `DynamicObjectParser`, then you can rename the flag and just check it there, without calling other methods.

exaV replied:
IMO `Dynamic` is not a serialisation format in the traditional sense, therefore I think it is fine to have options which will not play nice with the `DynamicObjectParser `. I will still set the default value of the option to encode null as `null`, so parsing the output of `DynamicObjectSerializer` should work by default.

EdwarDDay commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,182 @@
+ * * Enums are serialized as the value of `@SerialName` if present or their name, in that order. Ordinals are not supported.
+ * * There is no analogue to DynamicObjectParsers' useArrayPolymorphism flag
+ *
+ * @param encodeDefaults if true any property which equals exactly its default value will be omitted
```
if `encodeDefaults` is true, the default values should be encoded and not omitted.
```suggestion
 * @param encodeDefaults if false any property which equals exactly its default value will be omitted
```

EdwarDDay commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,182 @@
+ */
+class DynamicObjectSerializer(
+    private val encodeDefaults: Boolean = true,
+    private val encodeNullAsUndefined: Boolean = false,
```
Maybe create a configuration class, like `JsonConfiguration`.

exaV replied:
It seems like an overkill to me. I'd rather discuss whether we need the options in the first place. It's hard for me to imagine a use case for `encodeDefaults` (in the context of DynamicObjectSerializer) and while I have a use case for `encodeNullAsUndefined` it is also a bit out there.

My use-case for `encodeNullAsUndefined`: I want to pass some data to a table component (e.g. https://material-table.com/#/). I know (that for some business reason) some property will always be null, therefore I use `encodeNullAsUndefined ` to omit nulls which means there will be no column for that property in the table (instead of an empty one).

sandwwraith replied:
I think it is fine to have two separate flags. Maybe it's even ok to put full `JsonConfiguration` itself here, see https://github.com/Kotlin/kotlinx.serialization/pull/694/files#r376207786 , but I do not insist.

sandwwraith commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,182 @@
+ * * Map keys must be of primitive or enum type
+ * * Custom serializers are not supported
+ * * Enums are serialized as the value of `@SerialName` if present or their name, in that order. Ordinals are not supported.
+ * * There is no analogue to DynamicObjectParsers' useArrayPolymorphism flag
```
I'd rephrase this to 'currently does not support polymorphism'. It seems it doesn't work even with array polymorphism now. I have some thoughts how to fix that, so you can leave that to me.

sandwwraith commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,182 @@
+            StructureKind.MAP -> WriteMode.MAP
+            is PrimitiveKind, UnionKind.ENUM_KIND -> {
+                // the two cases are handled in DynamicObjectSerializer. But compiler does not know
+                throw IllegalArgumentException("DynamicObjectSerializer does not support serializing singular values of primitive or enum type.")
```
```suggestion
                error("DynamicObjectSerializer does not support serialization of singular primitive values or enum types.")
```

sandwwraith commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,182 @@
+ * * Custom serializers are not supported
+ * * Enums are serialized as the value of `@SerialName` if present or their name, in that order. Ordinals are not supported.
+ * * There is no analogue to DynamicObjectParsers' useArrayPolymorphism flag
+ *
```
It also may be useful to put code sample here from runtime_usage.md, how it is done for JSON: https://github.com/kotlin/kotlinx.serialization/blob/85c0b12df4d7656a81bd4997b1b0a21ab91da35b/runtime/commonMain/src/kotlinx/serialization/json/Json.kt#L25

sandwwraith commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,182 @@
+            }
+        }
+
+    override fun toString(): String {
```
Since it is useful only as a debug information while developing, I suggest removing this before merging.

exaV commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,182 @@
+ *
+ * Limitations:
+ * * Map keys must be of primitive or enum type
+ * * Custom serializers are not supported
```
Thanks, I did not realise. I added a `DynamicPrimitiveEncoder : ElementValueEncoder` which takes care of serialising the primitive and enum cases. Now we support custom serialisers in all cases.

exaV commented:
I also discovered that Longs were not serialised correctly. I now convert them to double before assigning to dynamic. This works for longs where abs(value) <= MAX_SAFE_INTEGER. Larger longs will be converted with a loss of precision if `JsonConfiguration.isLenient` is set, otherwise an exception is thrown.

exaV commented:
I rebased onto kotlinx/serialisation:dev, updated my api usage (replace deprecated stuff i used and so on) and squashed my commits.
I also discovered and fixed an edge case where Double is used as a map key with non-integer values.

sandwwraith commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
````diff
@@ -0,0 +1,252 @@
+ *  val plainJS: dynamic = DynamicObjectSerializer().serialize(DataWrapper.serializer(), wrapper)
+ * ```
+ *
+ * @param encodeNullAsUndefined if true null properties will be omitted from the output
````
This param doc is obsolete now

exaV replied:
The parameter is still there as well. Is there a well suited parameter that's already part of JsonConfiguration we could use?

sandwwraith replied:
Indeed, my bad. There is no suitable parameter [yet], so you may leave it as is.

sandwwraith commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,252 @@
+
+        if (!configuration.isLenient && conversionHasLossOfPrecision) {
+            throw IllegalArgumentException(
+                "$value can't be deserialized to number due to a potential precision loss. " +
```
```suggestion
                "$value can't be serialized to number due to a potential precision loss. " +
```

sandwwraith commented on runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt:
```diff
@@ -0,0 +1,252 @@
+        if (!configuration.isLenient && conversionHasLossOfPrecision) {
+            throw IllegalArgumentException(
+                "$value can't be deserialized to number due to a potential precision loss. " +
+                        "Use the JsonConfiguration option isLenient to serialise anyway"
```
```suggestion
                        "Use the JsonConfiguration option isLenient to serialize anyway"
```

sandwwraith commented:
Thank you once again for your help and work!
